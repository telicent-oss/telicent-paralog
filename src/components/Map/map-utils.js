import * as turf from "@turf/turf";
import { isEmpty } from "lodash";

const filterEmptyElements = (element) => !isEmpty(element);

export const generatePointAssetFeatures = (
  assets,
  dependencies,
  selectedElements,
) => {
  const pointAssets = assets
    .filter((asset) => asset.lat && asset.lng)
    .map((asset) => {
      return asset.createPointAsset();
    })
    .filter(filterEmptyElements);

  const pointAssetDependencies = dependencies
    .map((dependency) => {
      return dependency.createLineFeature(assets, selectedElements);
    })
    .filter(filterEmptyElements);
  return [...pointAssetDependencies, ...pointAssets];
};

export const generateLinearAssetFeatures = (assets, selectedElements) => {
  const linearAssets = assets
    .filter((asset) => !isEmpty(asset.geometry))
    .map((asset) => asset.createLinearAsset(selectedElements))
    .filter(filterEmptyElements);

  return linearAssets;
};

export const fitMultiToBounds = (map, selectedElements, assets) => {
  if (isEmpty(selectedElements) || isEmpty(assets)) return;

  const lngLats = selectedElements
    .filter((element) => element?.isPointAsset && element?.hasLatLng)
    .map((element) => [element.lng, element.lat]);
  const pointAssetMPFeature = turf.multiPoint(lngLats);

  const linearAssetGeometry = selectedElements
    .filter((element) => element?.isLinearAsset)
    .map((element) => element.createSegmentCoords());
  const linearAssetMLSFeature = turf.multiLineString(linearAssetGeometry);

  const dependencyGeometry = selectedElements
    .filter((element) => element?.isDependency)
    .map((element) => element.generateCoordinates(assets));
  const dependencyMLSFeature = turf.multiLineString(dependencyGeometry);

  if (
    isEmpty(pointAssetMPFeature.geometry.coordinates) &&
    isEmpty(linearAssetMLSFeature.geometry.coordinates) &&
    isEmpty(dependencyMLSFeature.geometry.coordinates)
  )
    return;

  const collection = turf.featureCollection([
    pointAssetMPFeature,
    linearAssetMLSFeature,
    dependencyMLSFeature,
  ]);
  const bbox = turf.bbox(collection);

  map.fitBounds(bbox, {
    padding: { top: 15, bottom: 25, left: 15, right: 55 },
  });
};

const fitLineStringToBounds = (map, geometry) => {
  if (isEmpty(geometry)) return;

  const lineString = turf.lineString(geometry);
  const bbox = turf.bbox(lineString);

  map.fitBounds(bbox, {
    padding: { top: 10, bottom: 25, left: 15, right: 5 },
  });
};

export const fitToBounds = (map, selectedElement, assets) => {
  if (selectedElement?.isPointAsset && selectedElement?.hasLatLng) {
    map.flyTo({ center: [selectedElement.lng, selectedElement.lat], zoom: 12 });
    return;
  }

  if (selectedElement?.isLinearAsset) {
    const geometry = selectedElement.createSegmentCoords();
    fitLineStringToBounds(map, geometry);
    return;
  }

  if (selectedElement?.isDependency && !isEmpty(assets)) {
    const geometry = selectedElement.generateCoordinates(assets);
    fitLineStringToBounds(map, geometry);
    return;
  }
};
