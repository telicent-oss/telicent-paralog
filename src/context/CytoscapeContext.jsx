import { isEmpty } from "lodash";
import React, { createContext, useCallback, useRef } from "react";

import { useLocalStorage } from "hooks";
import { getUniqueElements } from "utils";

export const CytoscapeContext = createContext();

export const CytoscapeProvider = ({ children }) => {
  const cyRef = useRef({});
  const [layout, setLayout] = useLocalStorage("graphLayout", "cola");

  const moveTo = ({ areaSelect, cachedElements, selectedElements }) => {
    if (!cyRef.current) return;
    const padding = 20;
    const selected = areaSelect
      ? selectedElements
      : getUniqueElements([...cachedElements, ...selectedElements]);

    const elements = cyRef.current.elements().filter((element) => {
      return selected.some((selectedElement) => {
        const data = element.data("element");
        return selectedElement.uri === data.uri;
      });
    });

    fit(elements, padding);
  };

  const fit = (elements, padding) => {
    if (!cyRef.current) return;
    cyRef.current.fit(elements, padding);
  };

  const updateLayout = (layout) => {
    setLayout(layout);
  };

  const resize = () => {
    if (isEmpty(cyRef?.current)) return;
    cyRef.current.resize();
  };

  const runLayout = useCallback(() => {
    if (isEmpty(cyRef?.current) || cyRef.current?.destroyed()) return;
    const cylayout = cyRef.current.layout({ name: layout });
    cylayout.run();
  }, [cyRef, layout]);

  return (
    <CytoscapeContext.Provider
      value={{ cyRef, layout, fit, moveTo, resize, runLayout, updateLayout }}
    >
      {children}
    </CytoscapeContext.Provider>
  );
};
