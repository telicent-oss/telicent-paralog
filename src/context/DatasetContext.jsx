import { createContext, useState } from "react";

const DatasetContext = createContext();

const DatasetProvider = ({ children }) => {
  const [selectedFloodAreas, setSelectedFloodAreas] = useState([]);
  const [expandedFloodAreas, setExpandedFloodAreas] = useState([]);
  const [selectedAssets, setSelectedAssets] = useState([]);
  const [expandedAssets, setExpandedAssets] = useState([]);

  const addSelectedFloodAreas = (polygonUris) => {
    setSelectedFloodAreas(polygonUris);
  };

  const addSelectedAssets = (assets) => {
    console.log({ assets });
    setSelectedAssets(assets);
  };

  const removeSelectedAsset = (selectedAsset) => {
    setSelectedAssets((assets) => assets.filter((asset) => asset !== selectedAsset));
  };

  const toggleExpandedAssets = (selectedGroup) => {
    const index = expandedAssets.findIndex((ontologyGroup) => ontologyGroup === selectedGroup);
    if (index === -1) {
      setExpandedAssets([...expandedAssets, selectedGroup]);
      return;
    }
    const filteredGroups = expandedAssets.filter(
      (ontologyGroup) => ontologyGroup !== selectedGroup
    );
    setExpandedAssets(filteredGroups);
  };

  const updateExpandedFloodAreas = (groups) => {
    setExpandedFloodAreas(groups);
  };

  return (
    <DatasetContext.Provider
      value={{
        expandedAssets,
        expandedFloodAreas,
        selectedFloodAreas,
        selectedAssets,
        addSelectedAssets,
        addSelectedFloodAreas,
        removeSelectedAsset,
        toggleExpandedAssets,
        updateExpandedFloodAreas,
      }}
    >
      {children}
    </DatasetContext.Provider>
  );
};

export { DatasetContext, DatasetProvider };
