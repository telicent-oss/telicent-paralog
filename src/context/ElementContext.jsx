import React, { useCallback, useReducer, useState } from "react";

import elementsReducer, {
  ADD_ASSETS,
  ADD_DEPENDENCIES,
  AREA_SELECTION,
  CLEAR_SELECTED,
  DISMISS_ERROR,
  FILTER_SELECTED_ELEMENTS,
  INITIAL_STATE,
  MULTI_SELECT_ELEMENTS,
  REMOVE_ELEMENTS_BY_TYPE,
  RESET,
  SELECT_ELEMENT,
  UPDATE_ERRORS,
} from "./elements-reducer";

export const ElementsContext = React.createContext();

export const ElementsProvider = ({ children }) => {
  const [state, dispatch] = useReducer(elementsReducer, INITIAL_STATE);
  const [selectedTimeline, setSelectedTimeline] = useState(null);

  const onFloodTimelineSelect = (selected) => {
    if (selected !== selectedTimeline) {
      setSelectedTimeline(selected);
    }
    return;
  };

  const closeTimelinePanel = () => {
    setSelectedTimeline(null);
  };

  const {
    assets,
    dependencies,
    errors,
    selectedElements,
    maxAssetCriticality,
    maxAssetTotalCxns,
    assetCriticalityColorScale,
    cxnCriticalityColorScale,
    totalCxnsColorScale,
  } = state;

  const filterSelectedElements = useCallback(() => {
    dispatch({ type: FILTER_SELECTED_ELEMENTS });
  }, []);

  const addElements = useCallback(
    (assets, dependencies) => {
      if (!Array.isArray(assets) || !Array.isArray(dependencies)) return;
      dispatch({ type: ADD_ASSETS, assets });
      dispatch({ type: ADD_DEPENDENCIES, dependencies });
      filterSelectedElements();
    },
    [filterSelectedElements]
  );

  const removeElementsByType = useCallback(
    (typeUri) => {
      if (!typeUri) return;
      dispatch({ type: REMOVE_ELEMENTS_BY_TYPE, typeUri });
      filterSelectedElements();
    },
    [filterSelectedElements]
  );

  const reset = useCallback(() => {
    dispatch({ type: RESET });
  }, []);

  const onElementClick = useCallback((multiSelect, selectedElements) => {
    if (multiSelect) {
      dispatch({ type: MULTI_SELECT_ELEMENTS, selectedElements });
      return;
    }
    dispatch({ type: SELECT_ELEMENT, selectedElements });
  }, []);

  const onAreaSelect = useCallback((selectedElements) => {
    if (!Array.isArray(selectedElements)) return;
    dispatch({ type: AREA_SELECTION, selectedElements });
  }, []);

  const updateErrorNotifications = useCallback((msg) => {
    dispatch({ type: UPDATE_ERRORS, error: msg });
  }, []);

  const dismissErrorNotification = (error) => {
    dispatch({ type: DISMISS_ERROR, error });
  };

  const clearSelectedElements = useCallback(() => {
    dispatch({ type: CLEAR_SELECTED });
  }, []);

  return (
    <ElementsContext.Provider
      value={{
        assets,
        dependencies,
        errors,
        selectedElements,
        maxAssetCriticality,
        maxAssetTotalCxns,
        assetCriticalityColorScale,
        cxnCriticalityColorScale,
        totalCxnsColorScale,
        addElements,
        clearSelectedElements,
        dismissErrorNotification,
        filterSelectedElements,
        onAreaSelect,
        onElementClick,
        reset,
        removeElementsByType,
        updateErrorNotifications,
        selectedTimeline,
        closeTimelinePanel,
        onFloodTimelineSelect,
      }}
    >
      {children}
    </ElementsContext.Provider>
  );
};
