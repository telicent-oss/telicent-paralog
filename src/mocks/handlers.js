import { http } from "msw";

import {
  createOntologyServiceEndpoint,
  createParalogEndpoint,
} from "api/combined";
import {
  ASSESSMENTS_ASSETS_ENDPOINT,
  ASSESSMENTS_DEPENDENCIES_ENDPOINT,
} from "constants/endpoints";

import * as resolvers from "./resolvers";

export const handlers = [
  http.get(createParalogEndpoint("assessments"), resolvers.assessments),
  http.get(
    createParalogEndpoint("assessments/asset-types"),
    resolvers.assetTypes,
  ),
  http.get(ASSESSMENTS_ASSETS_ENDPOINT, resolvers.mockEmptyResponse),
  http.get(ASSESSMENTS_DEPENDENCIES_ENDPOINT, resolvers.mockEmptyResponse),
  http.get(createParalogEndpoint("asset"), resolvers.asset),
  http.get(createParalogEndpoint("asset/dependents"), resolvers.dependents),
  http.get(createParalogEndpoint("asset/providers"), resolvers.providers),
  http.get(createParalogEndpoint("asset/residents"), resolvers.residents),
  http.get(createParalogEndpoint("flood-watch-areas"), resolvers.allFloodAreas),
  http.get(
    createParalogEndpoint("flood-watch-areas/polygon"),
    resolvers.floodAreaPolygons,
  ),
  http.get(
    createParalogEndpoint("person/residences"),
    resolvers.personResidences,
  ),
  http.get(createParalogEndpoint("ontology/class"), resolvers.ontologyClass),
  http.get(
    createOntologyServiceEndpoint("ontology/query"),
    resolvers.ontologyStyles,
  ),
  http.all("*", (req, res, ctx) => {
    console.warn("[MSW] Unhandled request to:", req.url.href);
    return res(ctx.status(500));
  }),
];
