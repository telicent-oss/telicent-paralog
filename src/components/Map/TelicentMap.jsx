import React, { useCallback, useContext, useEffect, useMemo, useState } from "react";
import Map, { Layer, Source, ScaleControl, useMap, AttributionControl } from "react-map-gl/maplibre";
import { ErrorBoundary } from "react-error-boundary";

import { CytoscapeContext, DatasetContext, ElementsContext } from "context";
import {
  useFloodAreaPolygons,
  useFloodMonitoringStations,
  useMapInteractions,
  useLocalStorage,
  useBuildingsEpcRating,
} from "hooks";
import { ErrorFallback, FloatingPanel, Modal } from "lib";

import MapToolbar from "./MapToolbar/MapToolbar";
import PointerCoordinates from "./PointerCoords";
import FloodMonitoringStations from "./FloodMonitoringStations";
import FloodWarningWidget from "./FloodAreaWidget";
import FloodZones from "./FloodZones";
import PointAssets from "./PointAssets";
import BuildingsEpcRating from "./BuildingsEpcRating";

import { FLOOD_AREA_LAYERS, heatmap, LINEAR_ASSET_LAYER } from "./layers";
import { generateLinearAssetFeatures } from "./map-utils";
import { getMapStyles } from "./mapStyles";

import "./map.css";

export const GEOJSON = "geojson";
export const FEATURE_COLLECTION = "FeatureCollection";
const VIEWSTATE = {
  latitude: 50.66206632912732,
  longitude: -1.3480234953335598,
  zoom: 9,
};
const HEAT_RADIUS = 1000;

const TelicentMap = () => {
  const { telicentMap: map } = useMap();

  const { moveTo } = useContext(CytoscapeContext);
  const { selectedFloodAreas } = useContext(DatasetContext);
  const { assets, dependencies, selectedElements, onElementClick, onAreaSelect } =
    useContext(ElementsContext);

  const { polygonFeatures: floodAreas, isLoading: areFloodAreasLoading } =
    useFloodAreaPolygons(selectedFloodAreas);
  const { interactiveLayers, selectedFloodZones, handleOnClick } = useMapInteractions({
    map,
    assets,
    dependencies,
    selectedElements,
    onElementClick,
    onAreaSelect,
    moveTo,
  });
  const mapStyles = useMemo(() => getMapStyles(), []);

  const [mapStyle, setMapStyle] = useLocalStorage("mapStyle", mapStyles[0]);
  const {
    query,
    menuItem: monitoringStationLayerItem,
    showStations,
  } = useFloodMonitoringStations();
  const {
    query: buildingsEpcQuery,
    menuItem: buildingsEpcLayerItem,
    showBuildings,
  } = useBuildingsEpcRating();

  const [cursor, setCursor] = useState("auto");
  const [heatmapRadius, setHeatmapRadius] = useState(10);
  const [showPointerCoords, setShowPointerCoords] = useState(false);

  const [linearAssets, setLinearAssets] = useState([]);
  const [mousePosition, setMousePosition] = useState(undefined);

  // The order of the array is the order in which the features will appear in the map.
  // index 0 being the lowest level
  const sources = useMemo(
    () => [
      {
        id: "flood-areas",
        features: floodAreas,
        layers: FLOOD_AREA_LAYERS,
      },
      { id: "linear-assets", features: linearAssets, layers: [LINEAR_ASSET_LAYER] },
    ],
    [linearAssets, floodAreas]
  );

  useEffect(() => {
    if (!map) return
    map.on('styledata', () => {
      const layer = map.getLayer('gl-draw-lines.hot');
      if (layer) {
        map.setPaintProperty('gl-draw-lines.hot', 'line-dasharray', ['literal', [2, 4]]);
      }
    });
  }, [map])

  useEffect(() => {
    const isStyleDefined = mapStyles.some((style) => style.id === mapStyle.id);
    if (!isStyleDefined) setMapStyle(mapStyles[0]);
  }, [mapStyles, mapStyle, setMapStyle]);

  useEffect(() => {
    const linearAssets = generateLinearAssetFeatures(assets, selectedElements);
    setLinearAssets(linearAssets);
  }, [assets, selectedElements]);

  const handleOnMouseMove = (event) => {
    const { lngLat } = event;
    setMousePosition(lngLat);
  };

  const resetCursor = () => {
    setCursor("auto");
  };

  const handleOnZoom = useCallback((event) => {
    if (!map) return;
    const { _pixelPerMeter } = event.target.transform;
    let radius = HEAT_RADIUS * _pixelPerMeter;
    if (radius < 1) radius = 1;
    map.getMap().setPaintProperty(heatmap.id, "heatmap-radius", radius);
    setHeatmapRadius(radius);
  }, [map]);

  const togglePointerCoords = () => {
    setShowPointerCoords((prev) => !prev);
  };

  const generateSources = (source) => (
    <Source
      key={source.id}
      id={source.id}
      type={GEOJSON}
      data={{ type: FEATURE_COLLECTION, features: source.features }}
      generateId
    >
      {source.layers.map((layer) => (
        <Layer key={layer.id} {...layer} />
      ))}
    </Source>
  );

  return (
    <ErrorBoundary FallbackComponent={ErrorFallback}>
      <div className="relative w-full">
        <Map
          cursor={cursor}
          id="telicentmap"
          interactiveLayerIds={interactiveLayers}
          initialViewState={{ ...VIEWSTATE }}
          mapboxAccessToken="MapboxToken"
          mapStyle={mapStyle?.id}
          attributionControl={false}
          onClick={handleOnClick}
          onDragStart={() => setCursor("move")}
          onDragEnd={resetCursor}
          onMouseEnter={() => setCursor("pointer")}
          onMouseLeave={resetCursor}
          onMouseMove={handleOnMouseMove}
          boxZoom={false}
          onZoom={handleOnZoom}
          styleDiffing
        >
          {sources.map(generateSources)}
          <PointAssets
            map={map}
            assets={assets}
            dependencies={dependencies}
            selectedElements={selectedElements}
            onElementClick={onElementClick}
            moveTo={moveTo}
            onAreaSelect={onAreaSelect}
          />
          <BuildingsEpcRating map={map} query={buildingsEpcQuery} showBuildings={showBuildings} />
          <FloodMonitoringStations query={query} showStations={showStations} />
          <AttributionControl compact />
          <ScaleControl
            position="bottom-right"
            style={{
              position: "relative",
              backgroundColor: "#27272780",
              color: "#F5F5F5",
              borderColor: "#949494",
              fontFamily: "Figtree",
              letterSpacing: "1.5px",
              margin: 0,
              height: "22px",
            }}
          />
          <MapToolbar
            heatmapRadius={heatmapRadius}
            map={map}
            mapStyle={mapStyle}
            setMapStyle={setMapStyle}
            showPointerCoords={showPointerCoords}
            onPointerCoordsClick={togglePointerCoords}
            setCursor={setCursor}
            layerItems={[monitoringStationLayerItem, buildingsEpcLayerItem]}
          />
        </Map>
        <TopLeftPanel>
          <PointerCoordinates
            show={showPointerCoords}
            lat={mousePosition?.lat}
            lng={mousePosition?.lng}
          />
          <FloodZones selectedFloodZones={selectedFloodZones} />
        </TopLeftPanel>
        <FloodWarningWidget />
      </div>
      <Modal appElement="root" isOpen={areFloodAreasLoading} className="px-6 py-2 rounded-lg">
        <p>Adding flood areas to map</p>
      </Modal>
    </ErrorBoundary>
  );
};

export default TelicentMap;

const TopLeftPanel = ({ children }) => (
  <FloatingPanel position="top-0 left-0">{children}</FloatingPanel>
);
