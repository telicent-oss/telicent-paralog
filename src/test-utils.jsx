import { render, screen } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import { Provider as UseHttpProvider } from "use-http";
import { MapProvider } from "react-map-gl";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { DSProviders } from "@telicent-oss/ds";
import OntologyService from "@telicent-oss/ontologyservice";

import { createAssets, createDependencies } from "components/Dataset/dataset-utils";
import config from "config/app-config";
import { Dataset } from "./components";
import { CytoscapeProvider, ElementsContext, ElementsProvider } from "./context";

const user = userEvent.setup();

const ontologyService = new OntologyService(config.services.ontology, "/ontology");

export const clickSmallWindFarm = async () => {
  await user.click(screen.getByRole("button", { name: /Green grid/i }));
  await user.click(await screen.findByRole("checkbox", { name: "Energy [25]" }));
  expect(screen.getByRole("checkbox", { name: "Energy [25]" })).toBeChecked();
};

export const clickEnergyDataset = async () => {
  await user.click(await screen.findByRole("checkbox", { name: "Energy [25]" }));
  expect(screen.getByRole("checkbox", { name: "Energy [25]" })).toBeChecked();
};

export const clickTransportDataset = async () => {
  await user.click(await screen.findByRole("checkbox", { name: "Transport [44]" }));
};

export const clickMedicalDataset = async () => {
  await user.click(await screen.findByRole("checkbox", { name: "Medical [32]" }));
};

export const expandPanel = async (panelName) =>
  await user.click(screen.getByRole("button", { name: panelName }));

export const AssetBtn = ({ label, assets, event, onElementClick }) => (
  <button
    onClick={() => {
      onElementClick(
        event,
        assets.find((asset) => asset.label === label)
      );
    }}
  >
    {label}
  </button>
);

export const CxnBtn = ({ label, connections, event, onElementClick }) => (
  <button
    onClick={() => {
      onElementClick(
        event,
        connections.find((cxn) => cxn.label === label)
      );
    }}
  >
    {label}
  </button>
);

export const PanelProviders = ({ children, elementsProviderValue }) => (
  <UseHttpProvider options={{ cacheLife: 0, cachePolicy: "no-cache" }}>
    <ElementsProvider value={elementsProviderValue}>{children}</ElementsProvider>
  </UseHttpProvider>
);

const LoadDataWrapper = ({ testComponent, children }) => (
  <MapProvider>
    <CytoscapeProvider>
      <ElementsProvider>
        <ElementsContext.Consumer>
          {({ assets, connections, selectedElements, onElementClick }) => {
            return (
              <>
                {testComponent && testComponent({ assets, connections, onElementClick })}
                <div id="all-assets">
                  {assets.map((asset) => (
                    <p id="asset" key={asset.id}>
                      {asset.id}
                    </p>
                  ))}
                </div>
                <div id="all-connections">
                  {connections.map((cxn) => (
                    <p id="cxn" key={cxn.id}>
                      {cxn.id}
                    </p>
                  ))}
                </div>
                <div id="selected-elements">
                  {selectedElements.map((selectedElement) => (
                    <p id="selected-element" key={selectedElement.id}>
                      {selectedElement.id}
                    </p>
                  ))}
                </div>
                <Dataset />
                {children}
              </>
            );
          }}
        </ElementsContext.Consumer>
      </ElementsProvider>
    </CytoscapeProvider>
  </MapProvider>
);

export const renderTestComponent = (ui, options) => {
  return {
    user: userEvent.setup(),
    ...render(ui, {
      wrapper: ({ children }) => (
        <LoadDataWrapper testComponent={options?.testComponent}>{children}</LoadDataWrapper>
      ),
      ...options?.testingLibraryOptions,
    }),
  };
};

const createTestQueryClient = () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });
  return queryClient;
};
export const createQueryClientWrapper = () => {
  const queryClient = createTestQueryClient();
  return ({ children }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

export const renderWithQueryClient = (ui, options) => {
  const queryClient = createTestQueryClient();

  // https://github.com/testing-library/react-testing-library/issues/218#issuecomment-436730757
  const rendered = render(
    <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>,
    options
  );

  return {
    user,
    ...rendered,
    rerender: (ui, options) =>
      renderWithQueryClient(ui, { container: rendered.container, ...options }),
  };
};

export const setup = (ui) => {
  return {
    user,
    ...render(ui),
  };
};

export const getCreatedAssets = async (
  assets,
  ids,
  findIcon = jest.fn(),
  getAssetGeometry = jest.fn()
) => {
  const createdAssets = (await createAssets(assets, findIcon, getAssetGeometry)).filter((asset) =>
    ids.includes(asset.id)
  );
  return createdAssets;
};

export const getCreatedDependencies = (dependencies, ids) => {
  const createdDependencies = createDependencies(dependencies).filter((asset) =>
    ids.includes(asset.id)
  );
  return createdDependencies;
};

export const DSProvidersWrapper = ({ children }) => (
  <DSProviders ontologyService={ontologyService}>{children}</DSProviders>
);
