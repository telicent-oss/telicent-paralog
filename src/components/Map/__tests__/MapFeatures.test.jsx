import React from "react";
import { vi } from "vitest";

import { CytoscapeProvider, DatasetContext, ElementsContext } from "context";
import { DSProvidersWrapper, getCreatedAssets, renderWithQueryClient } from "test-utils";
import { T020_SEGMENTS, T034_SEGMENTS, TRANSPORT_ASSETS } from "mocks";

import TelicentMap from "../TelicentMap";
import * as mapboxFeatures from "../map-utils";

const MapComponent = ({
  assets = [],
  dependencies = [],
  selectedElements = [],
  selectedFloodAreas = [],
}) => (
  <DSProvidersWrapper>
    <CytoscapeProvider>
      <ElementsContext.Provider
        value={{
          assets,
          dependencies,
          selectedElements,
          clearSelectedElements: vi.fn(),
          onElementClick: vi.fn(),
          closeTimelinePanel: vi.fn(),
        }}
      >
        <DatasetContext.Provider value={{ selectedFloodAreas }}>
          <TelicentMap />
        </DatasetContext.Provider>
      </ElementsContext.Provider>
    </CytoscapeProvider>
  </DSProvidersWrapper>
);

describe("Map component", () => {
  test("should generate empty features when there are no elements", async () => {
    const spyOnGenerateLinearAssetFeatures = vi.spyOn(
      mapboxFeatures,
      "generateLinearAssetFeatures"
    );
    renderWithQueryClient(<MapComponent />);

    expect(spyOnGenerateLinearAssetFeatures).toHaveReturnedWith([]);
  });

  test("should generate linear asset features", async () => {
    const spyOnGenerateLinearAssetFeatures = vi.spyOn(
      mapboxFeatures,
      "generateLinearAssetFeatures"
    );
    const createdAssets = await getCreatedAssets(
      TRANSPORT_ASSETS,
      ["T043"],
      vi.fn(),
      vi.fn().mockReturnValue(T034_SEGMENTS)
    );
    renderWithQueryClient(<MapComponent assets={createdAssets} />);

    expect(spyOnGenerateLinearAssetFeatures).toBeCalledTimes(1);
    expect(spyOnGenerateLinearAssetFeatures.mock.results[0].value).toHaveLength(1);
    expect(spyOnGenerateLinearAssetFeatures.mock.results[0].value).toMatchSnapshot();
  });

  test("should NOT generate linear asset features when previously selected assets are removed", async () => {
    const spyOnGenerateLinearAssetFeatures = vi.spyOn(
      mapboxFeatures,
      "generateLinearAssetFeatures"
    );
    const assets = await getCreatedAssets(
      TRANSPORT_ASSETS,
      ["T043"],
      vi.fn(),
      vi.fn().mockReturnValue(T034_SEGMENTS)
    );
    const { rerender } = renderWithQueryClient(
      <MapComponent assets={assets} selectedElements={assets} />
    );

    const results = spyOnGenerateLinearAssetFeatures.mock.results[0].value;
    expect(results).toHaveLength(1);
    expect(results[0].properties.selected).toBe(true);

    rerender(<MapComponent assets={[]} selectedElements={assets} />);
    expect(spyOnGenerateLinearAssetFeatures).toBeCalledTimes(2);
    expect(spyOnGenerateLinearAssetFeatures).toHaveReturnedWith([]);
  });

  test("should generate selected linear asset features for selected assets which exist", async () => {
    const spyOnGenerateLinearAssetFeatures = vi.spyOn(
      mapboxFeatures,
      "generateLinearAssetFeatures"
    );
    const assets = await getCreatedAssets(
      TRANSPORT_ASSETS,
      ["T043", "T020"],
      vi.fn(),
      vi
        .fn()
        .mockReturnValueOnce(T034_SEGMENTS)
        .mockReturnValueOnce(T020_SEGMENTS)
        .mockReturnValue([])
    );
    const { rerender } = renderWithQueryClient(
      <MapComponent assets={assets} selectedElements={assets} />
    );

    let results = spyOnGenerateLinearAssetFeatures.mock.results[0].value;

    expect(spyOnGenerateLinearAssetFeatures).toBeCalledTimes(1);
    expect(results).toHaveLength(2);
    expect(results[0].properties.selected).toBe(true);
    expect(results[1].properties.selected).toBe(true);

    const filteredAssets = assets.filter((asset) => {
      const typeFilter = "http://ies.example.com/ontology/ies#ShippingComplex";
      const isShippingComplex = asset.type === typeFilter;
      return isShippingComplex;
    });
    rerender(<MapComponent assets={filteredAssets} selectedElements={assets} />);
    results = spyOnGenerateLinearAssetFeatures.mock.results[1].value;

    expect(spyOnGenerateLinearAssetFeatures).toBeCalledTimes(2);
    expect(results).toHaveLength(1);
    expect(results[0].properties.selected).toBe(true);
  });
});
