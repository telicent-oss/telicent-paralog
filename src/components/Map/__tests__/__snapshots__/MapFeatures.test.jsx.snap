// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Map component > should generate linear asset features 1`] = `
[
  {
    "geometry": {
      "coordinates": [
        [
          2.90952,
          1.35815,
        ],
        [
          3.58275,
          2.62053,
        ],
        [
          5.40952,
          5.35815,
        ],
        [
          6.91228,
          6.35815,
        ],
        [
          7.91228,
          7.35815,
        ],
        [
          -8.53063,
          8.35815,
        ],
      ],
      "type": "LineString",
    },
    "properties": {
      "criticality": 1234,
      "id": "T043",
      "lineColor": "#7C7C7C",
      "selected": false,
      "uri": "https://www.example.com/Instruments#T043",
    },
    "type": "Feature",
  },
]
`;
